File: spring/src/main/kotlin/one/digitalinnovation/gof/controller/ClientRestController.java
package one.digitalinnovation.gof.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import one.digitalinnovation.gof.model.Client;
import one.digitalinnovation.gof.service.ClientService;

/**
 * Esse {@link RestController} representa nossa <b>Facade</b>, pois abstrai toda
 * a complexidade de integrações (Banco de Dados H2 e API do ViaCEP) em uma
 * interface simples e coesa (API REST).
 *
 * @author falvojr
 */
@RestController
@RequestMapping("clientes")
public class ClientRestController {

	@Autowired
	private ClientService clientService;

	@GetMapping
	public ResponseEntity<Iterable<Client>> buscarTodos() {
		return ResponseEntity.ok(clientService.getAll());
	}

	@GetMapping("/{id}")
	public ResponseEntity<Client> buscarPorId(@PathVariable Long id) {
		return ResponseEntity.ok(clientService.getById(id));
	}

	@PostMapping
	public ResponseEntity<Client> inserir(@RequestBody Client cliente) {
		clientService.insert(cliente);
		return ResponseEntity.ok(cliente);
	}

	@PutMapping("/{id}")
	public ResponseEntity<Client> atualizar(@PathVariable Long id, @RequestBody Client client) {
		clientService.update(id, client);
		return ResponseEntity.ok(client);
	}

	@DeleteMapping("/{id}")
	public ResponseEntity<Void> deletar(@PathVariable Long id) {
		clientService.delete(id);
		return ResponseEntity.ok().build();
	}
}

File: spring/src/main/kotlin/one/digitalinnovation/gof/controller/ClientRestController.kt
package one.digitalinnovation.gof.controller

import one.digitalinnovation.gof.model.Client
import org.springframework.beans.factory.annotation.Autowired
import one.digitalinnovation.gof.service.ClientService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

/**
 * Esse [RestController] representa nossa **Facade**, pois abstrai toda
 * a complexidade de integrações (Banco de Dados H2 e API do ViaCEP) em uma
 * interface simples e coesa (API REST).
 *
 * @author falvojr
 */
@RestController
@RequestMapping("clients")
class ClientRestController {
    @Autowired
    private val clientService: ClientService? = null
    @GetMapping
    fun getAll(): ResponseEntity<Iterable<Client?>?> {
        return ResponseEntity.ok(
            clientService!!.getAll()
        )
    }

    @GetMapping("/{id}")
    fun getById(@PathVariable id: Long): ResponseEntity<Client> {
        return ResponseEntity.ok(clientService!!.getById(id))
    }

    @PostMapping
    fun insert(@RequestBody client: Client): ResponseEntity<Client> {
        clientService!!.insert(client)
        return ResponseEntity.ok(client)
    }

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @RequestBody client: Client): ResponseEntity<Client> {
        clientService!!.update(id, client)
        return ResponseEntity.ok(client)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): ResponseEntity<Void> {
        clientService!!.delete(id)
        return ResponseEntity.ok().build()
    }
}